core/lib/zksync_core/src/state_keeper/batch_executor/main_executor.rs

```diff
@@ -94,13 +84,22 @@ impl BatchExecutor for MainBatchExecutor {
         };
         let upload_witness_inputs_to_gcs = self.upload_witness_inputs_to_gcs;
 
+        let cached_storage = self.cached_storage.clone();
+        let stop_receiver = stop_receiver.clone();
         let handle = tokio::task::spawn_blocking(move || {
-            executor.run(
-                secondary_storage,
-                l1_batch_params,
-                system_env,
-                upload_witness_inputs_to_gcs,
-            )
+            let rt_handle = Handle::current();
+            let mut cached_storage = rt_handle.block_on(cached_storage.write());
```

Comment:
- Nit: If you only use `RwLock::write()`, it makes sense to replace it with `Mutex`.
- In the general case, this would look like a *suspicious* use of `RwLock`: you lock it exclusively for an extended period of time. AFAIU, it works right now because L1 batches are processed strictly sequentially, but in general reducing the lock time to a minimum (like in the approach I've outlined in another comment) is always preferable.

Reply:
Right, I think I was confused by lifetimes here and presumed that the guard gets dropped after `access_storage`. So your other comment about explicit lifetime makes a lot of sense!